### PR TITLE
fix: User cannot move resource to class wherte resource is

### PR DESCRIPTION
**Pull request by bartlomiejmarszal**
User cannot move resource to this same class. 

1. Select resource and click `Move To`
2. Select Class where item is located
3. Observe error message `Unable to move the resources`

Screenshot 2024-07-02 at 11 51 22

https://github.com/oat-sa/tao-core/assets/16231681/0b8cebf8-081f-411f-972e-cd8a18e22410

### CODE REVIEW

File: actions/class.RdfController.php [hunk 1]
```diff
@@ -972,12 +972,22 @@ public function moveResource()
 
             try {
                 $this->validateCsrf();
+                $this->validateDestinationClass($destinationClassUri, $this->getResourceParentClass($id));
```
Comment (pnal) [inline actions/class.RdfController.php:975]: the methods called below throws some InvalidArgumentException but not catched here
```
$this->validateUri($id);
$this->validateInstanceRoot($id);
$this->validateMoveRequest();
```
Furthermore some of them validates params that you used above
for example:
```
protected function validateMoveRequest()
    {
        if (!$this->isRequestPost()) {
            throw new common_exception_MethodNotAllowed('Only POST method is allowed to move instances.');
        }

        if (!$this->hasRequestParameter('destinationClassUri')) {
            throw new InvalidArgumentException('Destination class must be specified');
        }

        $destinationClass = new core_kernel_classes_Class($this->getRequestParameter('destinationClassUri'));
        if (!$destinationClass->isClass()) {
            throw new InvalidArgumentException('Destination class must be a valid class');
        }
```

Comment (bartlomiejmarszal) [reply actions/class.RdfController.php:975]: But I catch this exception at most granular way.

File: actions/class.RestResource.php [hunk 2]
```diff
@@ -155,7 +155,8 @@ public function getAll()
                         $selectedUris,
                         $search,
                         $offset,
-                        $limit
+                        $limit,
+                        $this->getExcludedClasses()
```
Comment (Karol-Stelmaczonek) [inline actions/class.RestResource.php:159]: Can we mimic logic which is used for other parameters? In L:134 for example? Handle $this->hasGetParameter('exclude') outside you new function, and here just pass variable with return value from ` getExcludedClasses()` ?

File: actions/class.RestResource.php [hunk 3]
```diff
@@ -389,4 +390,41 @@ private function getAdvancedLogger(): LoggerInterface
...
+     * @return string
+     */
+    private function getExcludedClasses(): array
+    {
+        //Exclude exclude parameter resource parent class
+        if ($this->hasGetParameter('exclude')) {
+            $excludeParam = $this->getRawParameter('exclude');
+            /** @var core_kernel_classes_Resource $selectedResourceToExclude */
+            $selectedResourceToExclude = $this->getResource(reset($excludeParam));
+            if (!$selectedResourceToExclude->exists()) {
```
Comment (Karol-Stelmaczonek) [inline actions/class.RestResource.php:406]:
```suggestion
            if ($selectedResourceToExclude && !$selectedResourceToExclude->exists()) {
```

File: actions/class.RestResource.php [hunk 4]
```diff
@@ -389,4 +390,41 @@ private function getAdvancedLogger(): LoggerInterface
...
+
+            $types = $selectedResourceToExclude->getTypes();
+            if (count($types) !== 1) {
+                $this->getLogger()->error('Resource to exclude has to have only one parent class');
+                return [];
+            }
+
+            $resourceParentClass = reset($types);
+
+            if (!$resourceParentClass->isClass()) {
```
Comment (Karol-Stelmaczonek) [inline actions/class.RestResource.php:419]:
```suggestion
            if ($resourceParentClass && !$resourceParentClass->isClass()) {
```

File: models/classes/GenerisTreeFactory.php [hunk 5]
```diff
@@ -199,7 +203,10 @@ private function buildChildNodes(core_kernel_classes_Class $class, $subclasses)
 
         // subclasses
         foreach ($subclasses as $subclass) {
-            $children[] = $this->classToNode($subclass, $class);
+            $child = $this->classToNode($subclass, $class);
+            if (!in_array($child['attributes']['data-uri'], $this->excludedClasses)) {
```
Comment (Karol-Stelmaczonek) [inline models/classes/GenerisTreeFactory.php:207]:
```suggestion
            if (isset($child['attributes']['data-uri']) && !in_array($child['attributes']['data-uri'], $this->excludedClasses)) {
```

File: actions/class.RdfController.php [hunk 6]
```diff
@@ -1714,4 +1725,21 @@ private function getResourceTransfer(): ResourceTransferInterface
...
+    {
+        $types = $this->getResource($id)->getTypes();
+        if (count($types) !== 1) {
+            throw new InvalidArgumentException('Resource has no class or resource types are ambiguous');
+        }
+
+        /** @var core_kernel_classes_Class $type */
+        $type = reset($types);
+
+        if (!$type->isClass()) {
```
Comment (Karol-Stelmaczonek) [inline actions/class.RdfController.php:1739]:
```suggestion
        if ($type && !$type->isClass()) {
```

File: actions/class.RestResource.php [hunk 7]
```diff
@@ -132,6 +132,7 @@ public function getAll()
                 $search   = $this->hasRequestParameter('search') ? $this->getRawParameter('search') : '';
                 $limit    = $this->hasRequestParameter('limit') ? $this->getRequestParameter('limit') : 30;
                 $offset   = $this->hasRequestParameter('offset') ? $this->getRequestParameter('offset') : 0;
+                $excludedClasses = $this->getExcludedClasses();
```
Comment (Karol-Stelmaczonek) [inline actions/class.RestResource.php:135]:
```suggestion
                $excludedClasses = $this->hasGetParameter('exclude') ? $this->getExcludedClasses() : '';
```
And then remove if with `$this->hasGetParameter('exclude')` from `getExcludedClasses()`

Comment (bartlomiejmarszal) [reply actions/class.RestResource.php:135]: I'm not sure if this is more clean version.

**Comment by tikhanovichA** [comment]
The current implementation blocks moving ticket to child classes from the current class. For example if my resource now in class_1 then I will not able to move it into class_1_1.
Because of that, let's just show the error message for now
image

**Review by pnal — CHANGES_REQUESTED** [review]
It seems that fully BE solution will not work here

**Review by tikhanovichA — APPROVED** [review]
I think that we are ok with that BE solution with the error message
image

**Review Checklist**
- [ ] New code is covered by tests (if applicable)
- [ ] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [x] New code respects code style rules
- [x] New code respects best practices
- [x] Feature is working correctly on my local machine (if applicable)
- [x] Acceptance criteria are respected
- [x] Pull request title and description are meaningful
- [x] Pull request's target is not `master`
- [x] Commits are following [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/)
- [x] Commits messages are meaningful
- [x] Commits are atomic

**Review by pnal — APPROVED** [review]

**Pull request merged**